File: Problems/1268_Search_Suggestions_System.py
from typing import List

class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        curr = products
        res = []
        for i, c in enumerate(searchWord):
            nxt = []
            for name in curr:
                if i<len(name) and c == name[i]:
                    nxt.append(name)
            res.append(nxt[:3])
            curr = nxt
        return res

if __name__ == '__main__':
    sol = Solution()
    products = ["mobile","mouse","moneypot","monitor","mousepad"]
    searchWord = "mouse"
    print(sol.suggestedProducts(products, searchWord))

File: Problems/1268_Search_Suggestions_System.cpp
#include<vector>
#include<string>
#include<iostream>
#include<algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        sort(products.begin(), products.end());
        vector<vector<string>> res;
        vector<string> curr = products;

        for(int i=0; i<searchWord.length(); ++i){
            vector<string> nxt;
            for(auto word: curr){
                if(i<word.length() && searchWord[i]==word[i]){
                    nxt.push_back(word);
                }
            }
            curr = nxt;
            vector<string> three;
            for(int i=0; i<curr.size(); ++i){
                three.push_back(curr[i]);
                if(i==2){
                    break;
                }
            }
            res.push_back(three);
        }
        return res;
    }
};